lib/util/PriorityQueue.js

```diff
@@ -0,0 +1,107 @@
+"use strict";
+
+/**
+ * @template T
+ */
+class PriorityQueue {
+	/**
+	 * @param {function(T):number} priorityFn The function to calculate the item's priority. Lower values will be pushed to the front of the queue
+	 * @param {Iterable<T>=} items The initial elements.
+	 */
+	constructor(priorityFn, items) {
+		/** @type {{priority: number, item: T}[]} */
+		this.heap = [];
+		/** @type {function(T):number}*/
+		this.priorityFn = priorityFn;
+		for (let item of items) {
+			this.enqueue(item);
+		}
+	}
+
+	/**
+	 * Returns the number of elements in this queue.
+	 * @returns {number} The number of elements in this queue.
+	 */
+	get length() {
+		return this.heap.length;
+	}
+
+	/**
+	 * Appends the specified element to this queue.
+	 * @param {T} item The element to add.
+	 * @returns {void}
+	 */
+	enqueue(item) {
+		this.heap.push({ priority: this.priorityFn(item), item });
+		this._siftUp();
+	}
+
+	/**
+	 * Retrieves and removes the head of this queue.
+	 * @returns {T | undefined} The head of the queue of `undefined` if this queue is empty.
+	 */
+	dequeue() {
+		const lastIndex = this.length - 1;
+		if (lastIndex > 0) {
+			this._swap(0, lastIndex);
+		}
+		const poppedValue = this.heap.pop();
+		this._siftDown();
+		return poppedValue.item;
```

Comment:
Calling `dequeue` on an empty `PriorityQueue` would result in a `TypeError`. It would be better to change this line to:

```suggestion
		return poppedValue ? poppedValue.item : undefined;
```